components/style/matching.rs

```diff
@@ -736,70 +736,146 @@ pub trait MatchMethods : TNode {
+                                                   -> (Self::ConcreteRestyleDamage, RestyleResult)
+        where Ctx: StyleContext<'a>
+    {
+        // Here we optimise the case of the style changing but both the
+        // previous and the new styles having display: none. In this
+        // case, we can always optimize the traversal, regardless of the
+        // restyle hint.
+        let this_display = final_style.get_box().clone_display();
+        if this_display == display::T::none {
+            let old_display = data.style.as_ref().map(|old_style| {
+                old_style.get_box().clone_display()
+            });
+
+            // If display passed from none to something, then we need to reflow,
+            // otherwise, we don't do anything.
+            let damage = match old_display {
+                Some(display) if display == this_display => {
+                    Self::ConcreteRestyleDamage::empty()
+                }
+                _ => {
+                    Self::ConcreteRestyleDamage::rebuild_and_reflow()
+                }
+            };
+
+            debug!("Short-circuiting traversal: {:?} {:?} {:?}",
+                   this_display, old_display, damage);
+
+            data.style = Some(final_style);
+            return (damage, RestyleResult::Stop);
+        }
+
+        // Otherwise, we just compute the damage normally, and sum up the damage
+        // related to pseudo-elements.
+        let mut damage = {
+            let existing_style =
+                self.existing_style_for_restyle_damage(data.style.as_ref());
+
+            Self::ConcreteRestyleDamage::compute(existing_style, &final_style)
+        };
+
+        data.style = Some(final_style);
+
+        // FIXME(emilio): This is not pretty, and in the Gecko case means
+        // effectively comparing with the old computed values (given our style
+        // source is the old nsStyleContext).
+        //
+        // I... don't think any difference can arise from comparing with the old
+        // element restyle damage vs the new one, given that we're only summing
+        // the changes, and any change that we could miss would already have
+        // been caught by the parent's change. If for some reason I'm wrong on
+        // this, we'd have to compare computed values in Gecko too.
+        let existing_style =
+            self.existing_style_for_restyle_damage(data.style.as_ref());
+
+        let data_per_pseudo = &mut data.per_pseudo;
+        let new_style = data.style.as_ref();
+        debug_assert!(new_style.is_some());
+
+        <Self::ConcreteElement as MatchAttr>::Impl::each_eagerly_cascaded_pseudo_element(|pseudo| {
+            let applicable_declarations_for_this_pseudo =
+                applicable_declarations.per_pseudo.get(&pseudo).unwrap();
+
+            if !applicable_declarations_for_this_pseudo.is_empty() {
```

Comment:
nit: How about:

    if applicable_declarations_for_this_pseudo.is_empty() {
        return
    }

That way, we can avoid the extra layers of indentation below.